### PR TITLE
Kconfig: Add a new GUI configuration interface

### PR DESCRIPTION
**Pull request by ulfalizer**
Adds a new portable [Tkinter](https://docs.python.org/3/library/tk.html)-based configuration interface (run with `make/ninja guiconfig`).

Also piggybacks an update of Kconfiglib and the terminal menuconfig.

```
guiconfig: Add a graphical configuration interface

This is a graphical configuration interface written in Tkinter. Like
menuconfig.py, it supports showing all symbols (with invisible symbols
in red) and jumping directly to symbols. Symbol values can also be
changed directly from the jump-to dialog.

This interface should feel a lot smoother than menuconfig.py on Windows.

When single-menu mode is enabled, a single menu is shown at a time, like
in the terminal menuconfig. Only this mode distinguishes between symbols
defined with 'config' and symbols defined with 'menuconfig'.

Compatible with both Python 2 and Python 3. Has been tested on X11,
Windows, and macOS.

To avoid having to carry around a bunch of GIFs, the image data is
embedded in guiconfig.py. To use separate GIF files instead, change
_USE_EMBEDDED_IMAGES to False. The image files can be found in
https://github.com/ulfalizer/Kconfiglib/tree/screenshots/guiconfig.
```

```
doc: Add guiconfig documentation

Add documentation for the new GUI configuration interface in the
Application Development guide.

Also update all the images, touch up the language a bit, and tweak some
minor stuff:

 - Say *.conf instead of .conf to make it clearer that .conf isn't the
   literal name

 - Add a missing -D<board> argument when running cmake

 - Use figure:: instead of image::. Comes out left-aligned and with
   space between the images, which looks nicer.

 - Explain what '- -' and '-*-' means in the terminal menuconfig.

tkinter isn't included by default in many Python installations, despite
being part of the Python standard library, so also add the required
packages to the Development Environment Setup section of the manual. Add
a note to the Application Development section as well, to help
debugging.
```

```
menuconfig: Small Space/Enter improvement + fix for obscure bug

Update menuconfig (and Kconfiglib, just to sync) to upstream revision
7e0b9f7ae1, to get these commits in:

  menuconfig: Improve space/enter behavior slightly

  Space toggles value is possible, and enters menus otherwise. Enter
  works the other way around.

  Make this explicit in the code, which also fixes some corner cases,
  like space doing nothing on a y-selected menuconfig symbol.

  -------------------------------------------------------------------

  menuconfig: Fix display issue for unsatisfied-deps selected symbol
  with children

  A symbol with unsatisfied direct dependencies can end up with visible
  children in an implicit submenu if it is selected (though that
  generates a warning), so the optimization in _shown_nodes() isn't
  safe, and causes the child nodes to not be shown outside show-all
  mode.

  Just remove the optimization. Trying things out some more,
  everything's plenty fast enough anyway.

  Checking the direct dependencies of the parent instead would be safe.

The menu path now says "(Top)" instead of "(top menu)" too, which is a
bit more compact. Make the same change in genrest.py.
```

Screenshot:

guiconfig-jump-to

### CODE REVIEW

**Comment by ulfalizer** [comment]
I wonder why the CODEOWNERS check is still failing. Added an entry for `scripts/kconfig/`.

**Comment by ulfalizer** [comment]
@nashif 
Any way to get rid of that license check?

**Review by dbkinder — CHANGES_REQUESTED** [review]
Looks good, with a few tweaks...

File: doc/application/index.rst
Comment (ulfalizer) [reply doc/application/index.rst]: Typo. Fixed.

File: doc/application/index.rst
Comment (ulfalizer) [reply doc/application/index.rst]: Fixed

File: doc/application/index.rst
Comment (ulfalizer) [reply doc/application/index.rst]: Fixed

**Comment by ulfalizer** [comment]
@carlescufi 
Ping

**Comment by carlescufi** [comment]
Hey @ulfalizer sorry for the delay.

I just tried this on Linux:

```
$ ninja guiconfig
[0/1] cd /home/carles/src/zephyr/zephyr/build/zephyr/kconfig && /home/carles/bin/cmake/cmake-3.13.1-Linux-x86_64/bin/...f /usr/bin/python3 /home/carles/src/zephyr/zephyr/scripts/kconfig/guiconfig.py /home/carles/src/zephyr/zephyr/Kconfig
Traceback (most recent call last):
  File "/home/carles/src/zephyr/zephyr/scripts/kconfig/guiconfig.py", line 67, in <module>
    from tkinter import *
ModuleNotFoundError: No module named 'tkinter'
FAILED: CMakeFiles/guiconfig
cd /home/carles/src/zephyr/zephyr/build/zephyr/kconfig && /home/carles/bin/cmake/cmake-3.13.1-Linux-x86_64/bin/cmake -E env PYTHON_EXECUTABLE=/usr/bin/python3 srctree=/home/carles/src/zephyr/zephyr KERNELVERSION=0x10E6300 ZEPHYR_BASE=/home/carles/src/zephyr/zephyr KCONFIG_CONFIG=/home/carles/src/zephyr/zephyr/build/zephyr/.config ARCH=arm BOARD_DIR=/home/carles/src/zephyr/zephyr/boards/arm/reel_board SOC_DIR=/home/carles/src/zephyr/zephyr/soc CMAKE_BINARY_DIR=/home/carles/src/zephyr/zephyr/build ZEPHYR_TOOLCHAIN_VARIANT=zephyr ARCH_DIR=/home/carles/src/zephyr/zephyr/arch GENERATED_DTS_BOARD_CONF=/home/carles/src/zephyr/zephyr/build/zephyr/include/generated/generated_dts_board.conf /usr/bin/python3 /home/carles/src/zephyr/zephyr/scripts/kconfig/guiconfig.py /home/carles/src/zephyr/zephyr/Kconfig
ninja: build stopped: subcommand failed.

carles@zephyr-nrf ~/src/zephyr/zephyr/build (pr/15946)
$ python3
Python 3.6.7 (default, Oct 22 2018, 11:32:17)
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tkinter
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'tkinter'
```

**Review by carlescufi — CHANGES_REQUESTED** [review]
I see what happened now. You need to update the Linux getting started with:

`sudo apt install python3-tk`

**Comment by carlescufi** [comment]
This is pretty nifty @ulfalizer! I like it very much in general, though I do have a couple of objections:

- Checkboxes are disproportionately big, can you make the font size bigger to compensate?
- Text is too close to checkboxes, especially on macOS:

image

**Comment by ulfalizer** [comment]
> * Checkboxes are disproportionately big, can you make the font size bigger to compensate?

Played around with it a bit. I'm a bit scared of messing too much with the defaults though, since the default font varies by platform.

Maybe I drew the icons too large (16x16). Would take some time to tweak though, because there's some (crappy) pixel art involved. :P

Try playing around with this patch (`tweaked_font["size"] += 4`, etc.):

```diff
diff --git a/guiconfig.py b/guiconfig.py
index f682e10..619025a 100755
--- a/guiconfig.py
+++ b/guiconfig.py
@@ -361,17 +361,26 @@ def _fix_treeview_issues():
     # Fixes some Treeview issues
 
     global _treeview_rowheight
+    global tweaked_font  # Hack to prevent garbage collection
 
     style = ttk.Style()
 
+    tweaked_font = font.Font(font=style.lookup("Treeview", "font")).copy()
+    # Or, to set the font directly:
+    # tweaked_font = font.Font(font="Helvetica 12 bold italic")
+    #
+    # Can display default size with
+    # print(tweaked_font["size"])
+    tweaked_font["size"] += 4
+
     # The treeview rowheight isn't adjusted automatically on high-DPI displays,
     # so do it ourselves. The font will probably always be TkDefaultFont, but
     # play it safe and look it up.
 
-    _treeview_rowheight = font.Font(font=style.lookup("Treeview", "font")) \
-        .metrics("linespace") + 2
+    _treeview_rowheight = tweaked_font.metrics("linespace") + 2
 
-    style.configure("Treeview", rowheight=_treeview_rowheight)
+    style.configure("Treeview", rowheight=_treeview_rowheight,
+                    font=tweaked_font)
 
     # Work around regression in https://core.tcl.tk/tk/tktview?name=509cafafae,
     # which breaks tag background colors
```

> * Text is too close to checkboxes, especially on macOS:

Can't find a way to tweak it, unfortunately. Maybe it's just a bad standard, going on e.g. https://stackoverflow.com/questions/50319159/image-in-treeview-tkinter-library-in-python.

**Comment by ulfalizer** [comment]
Note to self:

Can query and modify the layout, but there doesn't seem to be a way to add padding. Might be hardcoded in `ttkTreeview.c`.

```python
    style.layout("Treeview.Item",
        [('Treeitem.padding',
          {'children': [('Treeitem.indicator', {'side': 'left', 'sticky': ''}),
                        ('Treeitem.image', {'side': 'left', 'sticky': ''}),
                        ('Treeitem.focus',
                         {'children': [('Treeitem.text',
                                        {'side': 'left', 'sticky': ''})],
                          'side': 'left',
                          'sticky': ''})],
           'sticky': 'nswe'})])
```

**Comment by ulfalizer** [comment]
@dbkinder @carlescufi 
Ping

**Comment by carlescufi** [comment]
> I see what happened now. You need to update the Linux getting started with:
> 
> `sudo apt install python3-tk`

@ulfalizer you are yet to address this

**Comment by ulfalizer** [comment]
> > I see what happened now. You need to update the Linux getting started with:
> > `sudo apt install python3-tk`
> 
> @ulfalizer you are yet to address this

Fixed. Also added a `tkinter` note to Application Development.

**Review by carlescufi — APPROVED** [review]
Nice work! This is a very useful addition. Once this is merged please announce this on the mailing list!

**Comment by ulfalizer** [comment]
@dbkinder @nashif 
Ping

The license failure is due to using ISC (same as for the rest of the Kconfiglib stuff).

File: scripts/kconfig/guiconfig.py
Comment (carlescufi) [inline scripts/kconfig/guiconfig.py]: Please change to  `Nordic Semiconductor ASA and Ulf Magnusson`

Comment (ulfalizer) [reply scripts/kconfig/guiconfig.py]: Fixed

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, thanks!

**Pull request merged**
